File: dlock-core/src/main/java/com/dlock/core/util/time/DateTimeProvider.java
package com.dlock.core.util.time;

import java.time.LocalDateTime;

/**
 * Managed date/time provider. With this testability is increased.
 *
 * @author Przemyslaw Malirz
 */
public interface DateTimeProvider {

    /**
     * Default singleton implementation.
     */
    DateTimeProvider SYSTEM = new DateTimeProvider() {
    };

    /**
     * Returns LocalDateTime.now(). It's not static so can be mocked / replaced by
     * any NOW provider.
     * The hardest part with NOW in unit tests is it's always different ;)
     */
    default LocalDateTime now() {
        return LocalDateTime.now();
    }
}

File: dlock-core/src/main/kotlin/com/dlock/util/time/DateTimeProvider.kt
package com.dlock.util.time

import java.time.LocalDateTime

/**
 * Managed date/time provider. With this testability is increased.
 *
 * @author Przemyslaw Malirz
 */
interface DateTimeProvider {

    companion object DefaultDateTimeProvider : DateTimeProvider

    /**
     * Returns LocalDateTime.now(). It's not static so can be mocked / replaced by any NOW provider.
     * The hardest part with NOW in unit tests is it's always different ;)
     */
    fun now(): LocalDateTime = LocalDateTime.now()

}